src/it/scala/org/apache/spark/sql/SFDataFrameWindowFunctionsSuite.scala

```diff
@@ -1,22 +1,810 @@
+//      val c1 = last(struct($"f2", $"f3")).over(w1) as "c1"
+//
+//      val input =
+//        """{"f1":1497820153720,"f2":[{"f20":"x","f21":0}],"f3":[{"f30":"x","f31":0}]}
+//          |{"f1":1497802179638}
+//          |{"f1":1497802189347}
+//          |{"f1":1497802189593}
+//          |{"f1":1497802189597}
+//          |{"f1":1497802189599}
+//          |{"f1":1497802192103}
+//          |{"f1":1497802193414}
+//          |{"f1":1497802193577}
+//          |{"f1":1497802193709}
+//          |{"f1":1497802202883}
+//          |{"f1":1497802203006}
+//          |{"f1":1497802203743}
+//          |{"f1":1497802203834}
+//          |{"f1":1497802203887}
+//          |{"f1":1497802203893}
+//          |{"f1":1497802203976}
+//          |{"f1":1497820168098}
+//          |""".stripMargin.split("\n").toSeq
+//
+//      import testImplicits._
+//
+//      assertSpilled(sparkContext, "select") {
+//        spark.read.schema(sampleSchema).json(input.toDS()).select(c0, c1).foreach { _ => () }
+//      }
+//    }
+//  }
+
+  test("SPARK-24575: Window functions inside WHERE and HAVING clauses") {
+    def checkAnalysisError(df: => DataFrame, clause: String): Unit = {
+      val thrownException = the[AnalysisException] thrownBy {
+        df.queryExecution.analyzed
+      }
+      assert(thrownException.message.contains(s"window functions inside $clause clause"))
+    }
+
+    checkAnalysisError(
+      testData2.select("a").where(rank().over(Window.orderBy($"b")) === 1), "WHERE")
+    checkAnalysisError(
+      testData2.where($"b" === 2 && rank().over(Window.orderBy($"b")) === 1), "WHERE")
+    checkAnalysisError(
+      testData2.groupBy($"a")
+        .agg(avg($"b").as("avgb"))
+        .where($"a" > $"avgb" && rank().over(Window.orderBy($"a")) === 1), "WHERE")
+    checkAnalysisError(
+      testData2.groupBy($"a")
+        .agg(max($"b").as("maxb"), sum($"b").as("sumb"))
+        .where(rank().over(Window.orderBy($"a")) === 1), "WHERE")
+    checkAnalysisError(
+      testData2.groupBy($"a")
+        .agg(max($"b").as("maxb"), sum($"b").as("sumb"))
+        .where($"sumb" === 5 && rank().over(Window.orderBy($"a")) === 1), "WHERE")
+
+    checkAnalysisError(sql("SELECT a FROM testData2 WHERE RANK() OVER(ORDER BY b) = 1"), "WHERE")
+    checkAnalysisError(
+      sql("SELECT * FROM testData2 WHERE b = 2 AND RANK() OVER(ORDER BY b) = 1"), "WHERE")
+    checkAnalysisError(
+      sql("SELECT * FROM testData2 GROUP BY a HAVING a > AVG(b) AND RANK() OVER(ORDER BY a) = 1"),
+      "HAVING")
+    checkAnalysisError(
+      sql("SELECT a, MAX(b), SUM(b) FROM testData2 GROUP BY a HAVING RANK() OVER(ORDER BY a) = 1"),
+      "HAVING")
+    checkAnalysisError(
+      sql(
+        s"""SELECT a, MAX(b)
+           |FROM testData2
+           |GROUP BY a
+           |HAVING SUM(b) = 5 AND RANK() OVER(ORDER BY a) = 1""".stripMargin),
+      "HAVING")
+  }
+
+//  test("window functions in multiple selects") {
```

Comment:
Again why is stuff commented out ?

Reply:
This is commented out because `stripAQEPlan(select.queryExecution.executedPlan)` is used, we don't support it yet. And it looks it is spark internal, it is not necessary for us.